Work around ICE in diagnostics for local super-universes missing `UniverseInfo`s

lqd opened the pull request:
In issue #114907, canonicalization of liveness dropck-outlives results (IIUC) encounters universes absent from the original query.  Some local universes [are created](https://github.com/lqd/rust/blob/f3a1bae88c617330b8956818da3cea256336c1cf/compiler/rustc_infer/src/infer/canonical/query_response.rs#L417-L425) for the mapping, but importantly, they won't have associated causes. 

These missing `UniverseInfo`s can be [needed](https://github.com/lqd/rust/blob/f3a1bae88c617330b8956818da3cea256336c1cf/compiler/rustc_borrowck/src/diagnostics/region_errors.rs#L376) during diagnostics, [causing the `IndexMap: key not found` ICE](https://github.com/lqd/rust/blob/d55522aad87c5605d7edd5dd4b37926e8b446117/compiler/rustc_borrowck/src/region_infer/mod.rs#L2252) seen in the issue.

This PR works around this by returning the suboptimal catch-all cause, to avoid the ICE. It does results in suboptimal diagnostics right now, but it's better than an ICE.

r? @matthewjasper. 

Let me know if there's a good easy-ish way to fix this, but I believe that for some of these erroneous cases and diagnostics, that inference/canonicalization/higher-ranked subtyping/etc may not behave exactly the same with the new trait solver? If that's the case then it'd probably be best to wait a bit more to do the correct fix.

Fixes #114907.

cc @aliemjay

lqd committed:
return default `UniverseInfo` cause in `RegionInferenceContext`

Query canonicalization can create local super-universes without causes,
creating ICEs when accessed during diagnostics.

lqd committed:
add test for issue 114907

matthewjasper reviewed (commented):
So, there's a choice between doing this and instead making sure that universe_causes has all universes that get created. I think this approach makes more sense because it's pretty hard to track down every code path that might add universes.

lqd committed:
remove dummy UniverseInfo causes from type checker `instantiate_canonical_with_fresh_inference_vars`

This was backfilling causes for the new universes that can be created by
the InferCtxt. We don't need to do that anymore: `other()` is the default when
there is no registered universe cause.

lqd committed:
remove dummy UniverseInfo causes from type checker `type_check`

This was pre-filling causes for universes that could already exist in
the InferCtxt. We don't need to do that anymore: `other()` is the default when
there is no registered universe cause.

lqd committed:
remove dummy UniverseInfo causes from type checker `fully_perform_op`

This was backfilling causes for new universes that may have been created
by an op, when there was no error info to use for improved
diagnostics. We don't need to do that anymore: `other()` is the default when
there is no registered universe cause.

lqd commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -2249,7 +2249,14 @@ impl<'tcx> RegionInferenceContext<'tcx> {
+        // #114907 where this happens via liveness and dropck outlives results.
+        // Therefore, we return a default value in case that happens, which should at worst emit a
+        // suboptimal error, instead of the ICE.
+        self.universe_causes.get(&universe).cloned().unwrap_or_else(|| UniverseInfo::other())
```
I've pushed a few commits to remove these. I think that's all of them.

bors commented:
:pushpin: Commit 10ef8d9bf7b9acb921b335b48b9b7a3294fe178b has been approved by `matthewjasper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=matthewjasper 10ef8d9bf7b9acb921b335b48b9b7a3294fe178b -->
<!-- homu: {"type":"Approved","sha":"10ef8d9bf7b9acb921b335b48b9b7a3294fe178b","approver":"matthewjasper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 10ef8d9bf7b9acb921b335b48b9b7a3294fe178b with merge 4b71f0303978581cb7444f0c493a0b35dab621e6...
<!-- homu: {"type":"BuildStarted","head_sha":"10ef8d9bf7b9acb921b335b48b9b7a3294fe178b","merge_sha":"4b71f0303978581cb7444f0c493a0b35dab621e6"} -->

rust-timer commented:
Finished benchmarking commit (4b71f0303978581cb7444f0c493a0b35dab621e6): [comparison URL](https://perf.rust-lang.org/compare.html?start=b30b535124aabf7835a09fe1c07e469711964011&end=4b71f0303978581cb7444f0c493a0b35dab621e6&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=b30b535124aabf7835a09fe1c07e469711964011&end=4b71f0303978581cb7444f0c493a0b35dab621e6&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b30b535124aabf7835a09fe1c07e469711964011&end=4b71f0303978581cb7444f0c493a0b35dab621e6&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1%  | [1.1%, 1.1%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 1.2%  | [1.1%, 1.3%]   | 4     |
| Improvements ✅ <br /> (primary)   | -1.6% | [-2.7%, -0.8%] | 4     |
| Improvements ✅ <br /> (secondary) | -1.2% | [-1.4%, -1.0%] | 2     |
| All ❌✅ (primary)                 | -1.1% | [-2.7%, 1.1%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b30b535124aabf7835a09fe1c07e469711964011&end=4b71f0303978581cb7444f0c493a0b35dab621e6&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.7%, -0.7%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.7% | [-0.7%, -0.7%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=b30b535124aabf7835a09fe1c07e469711964011&end=4b71f0303978581cb7444f0c493a0b35dab621e6&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 631.748s -> 629.802s (-0.31%)
**Artifact size**: 316.58 MiB -> 316.41 MiB (-0.05%)
